app/models/data_set.rb

```diff
@@ -62,18 +62,36 @@ def places_near(location, distance = nil, limit = nil, snac = nil)
   end
 
   def places_for_postcode(postcode, distance = nil, limit = nil)
-    location_data = MapitApi.location_for_postcode(postcode)
-    location = Point.new(latitude: location_data.lat, longitude: location_data.lon)
-    if service.location_match_type == "local_authority"
-      snac = MapitApi.extract_snac_from_mapit_response(location_data, service.local_authority_hierarchy_match_type)
-      if snac
-        places_near(location, distance, limit, snac)
-      else
-        []
-      end
-    else
-      places_near(location, distance, limit)
+    location_data = GdsApi.locations_api.coordinates_for_postcode(postcode)
+    raise GdsApi::HTTPNotFound, "Postcode exists, but has no location info" unless location_data
+
+    location = Point.new(latitude: location_data["latitude"], longitude: location_data["longitude"])
+    return places_near(location, distance, limit) if service.location_match_type == "nearest"
+
+    # TODO: This needs to be able to take into account an exact
+    # address so that we can handle split postcodes. This will
+    # involve some frontend work though. For now, do this to
+    # match the existing (not totally correct) behaviour
+    snac = appropriate_snac_for_postcode(postcode)
+    return [] unless snac
+
+    places_near(location, distance, limit, snac)
+  end
+
+  def appropriate_snac_for_postcode(postcode)
+    lcc = GdsApi.locations_api.local_custodian_code_for_postcode(postcode)
+    return nil if lcc.compact.empty?
+
+    response = GdsApi.local_links_manager.local_authority_by_custodian_code(lcc.first)
+    filtered_authorities = response.to_hash["local_authorities"].select do |la|
+      [service.local_authority_hierarchy_match_type, "unitary"].include?(la["tier"])
     end
+    snacs = filtered_authorities.collect { |la| la["snac"] }
```

Comment:
Minor: `map` would be more idiomatic than `collect` here but we can probably shorten the logic from line 89-92 with `filtered_authorities.first&.dig("snac")` which will return `nil` if `filtered_authorities` is an empty array or `snac` is `nil`.

Reply:
(change made)